### ISSUE TITLE
Embedding custom components

### ISSUE THREAD

**Issue by madhushan-t** [OvidijusParsiunas/deep-chat#18]
Hello!

Note: This is a fantastic and well-documented package, and it's exactly what developers starting out with AI chat applications need.

Now for my question...

I'm using this in an Angular project and everything works great, but I'm curious about the extensibility when we need to inject custom elements into the chat component. Here are a few use-cases:
- confirmation dialog
- email input text box
- user-feedback
- sharing

The response intercepters available today allow us to update the context in the response, but could we do the same with the text-box element?

Thanks in advance!

**Comment by OvidijusParsiunas** [comment]
Hi @madhushan-t, thankyou for raising the request!
Let me make sure I understand it correctly:

- For adding elements into the chat, are you referring for the ability for adding them beside the message responses? If that is correct - then it is good timing as I will be starting a new feature next week which will involve the ability to add buttons to AI responses which could potentially help you. But if you want to add your own custom elements, I am not sure if that is achievable given the current setup. If you can let me know more about those custom elements, I will be able to see the further possibility for their use. Other than that the only other option right now would be to fork the project and edit the source code yourself (which is not complicated).
- For the input text interceptor, do you want to edit the text that the user is typing? Could you elaborate a little bit more on how it would be used.

Thanks!

**Comment by madhushan-t** [comment]
Hi @OvidijusParsiunas,
- That's correct, I'm thinking of instances where an API may not only return text, but also some metadata that could be intercepted by this UI to render certain elements. Buttons are a great use-case! Another (more involved) example of this would be something similar to plotting graphs using ChatGPT plugins today:

<img width="286" alt="Screenshot 2023-09-22 at 8 23 42 PM" src="https://github.com/OvidijusParsiunas/deep-chat/assets/145786403/0354d4fc-e7a6-4600-992f-78ddb28c89c9">

But rather than dealing with 3rd-party plugins, I was thinking deep-chat could provide an API to inject custom elements underneath the AI chat response, similar to the image and text bubbles.

Or sometimes you may want unambiguous feedback to [make sure the user confirms something](https://www.bruinen.co/blog/confirmations-safely-take-actions-with-llms). Rather than have them type text, this could be a custom UI interaction. But I also see that this may be conflating traditional chat apps with what deep-chat is primarily intended for, which is text-based interactions.

As for my question about the text-box, I meant using your existing response-interceptor to implement the above use-case by modifying the html content, so instead of rendering the text response from the AI, the user can view a visualization, interact with a set of buttons, etc. Hope that makes sense?

Good luck with the upcoming work, and I'm looking forward to trying out those new features!

Best,
Madhu

**Comment by OvidijusParsiunas** [comment]
Thankyou for the explanation @madhushan-t. I experimented with this a little bit and see that messages can ultimately be replaced with your own custom html by instead of returning `{text: 'hello'}` you could instead reply with something like ```{html: `<button onclick="(() => {console.log('hello')})()">button</button>`}```. You could also then define your own custom elements like `<my-cystom-element></my-custom-element`. One issue that I found is that it is difficult to bind the javascript from the parent application to the button unless you are using global `window` which is an antipattern that I do not want to promote, so I'll try to see if I can do something about that.

Overall, I think it is achievable and I'll try to incorporate it into the buttons feature that I am planning to work on mid/late next week. (Reason for the delay is that I am about to release another project ([Active Table](https://github.com/OvidijusParsiunas/active-table)) and I am also making some changes in my career, so apologies for the wait and I am committed to adding this soon).

**Comment by madhushan-t** [comment]
Hi @OvidijusParsiunas, that's awesome. I think it will be a very valuable feature-addition.
And good luck with your other projects!

**Comment by OvidijusParsiunas** [comment]
Do not worry, the work for this will begin this week.
I am making other small updates to this project in the meanwhile to keep the repo active.

**Comment by OvidijusParsiunas** [comment]
Hey, thankyou for your patience, I am now going to begin work on this.

The goal will be to create a feature that would allow developers to render custom elements inside the chat.
Whilst the feature sounds relatively simple, the work for it will involve a multitude of steps - involving the refactoring of the component's codebase, enabling developers to bind their custom elements' events to their own functions and providing all of this in a cohesive API to enable simple usability. This will take a number of iterations to get right, so it may take some time for this feature to be released in the main Deep Chat package, but I will give you early access for it so you can validate against your usecases.

I will update this thread on the progress, thanks!

**Comment by OvidijusParsiunas** [comment]
Hi, I think I have the core functionality for this nailed down and you can test the code code described below using either of the following packages: `deep-chat-dev` and `deep-chat-react-dev` version `9.0.12`. These work exactly like the normal packages.

So the way custom elements are inserted is via the use of the new message content type property called `html`. This is placed inside the [Response](https://deepchat.dev/docs/connect#Response) type as follows:

```
Response {
  text?: string;
  files?: MessageFiles;
  html?: string;
  error?: string;
}
```

and the [MessageContent](https://deepchat.dev/docs/messages#MessageContent):

```
{role: MessageRole, text: string} | {role: MessageRole, file: MessageFile} | {role: MessageRole, file: string}
```

Example usage:
```
chatElementRef.initialMessages = [
    {
      role: 'user',
      html: '<button class="custom-button">Hello</button>',
    }
  ]
```

When it comes to adding styling and binding the elements to events, we will have to use the new `htmlClassUtilities` property. It works by mapping class names to events and stateful styles which has the following type:

```
type HTMLClassUtilities = {
  [className: string]: {events?: EventToFunction; styles?: StatefulStyles};
}

// the key here can be 'click', 'mousedown' and any other event that is accepted by `addEventListener`
type EventToFunction = {[K in keyof HTMLElementEventMap]?: () => void};
```
[StatefulStyles](https://deepchat.dev/docs/styles/#StatefulStyles) is already defined in the documentation.

Example use:

```
chatElementRef.htmlClassUtilities = {
    'custom-button': {
      events: {
        click: () => console.log('hello'),
      },
      styles: {
        default: {
          width: '200px',
        },
        hover: {
          backgroundColor: 'yellow',
        }
      }
    }
  }
```

To help with the element position you can also overwrite the message bubble styling using the [messageStyles](https://deepchat.dev/docs/messageStyles/) object's new `html` property where the new type will be:

```
MessageStyles {
  default?: MessageRoleStyles;
  html?: MessageElementsStyles;
  ... other properties
}
```
[MessageElementsStyles](https://deepchat.dev/docs/messageStyles/#MessageElementsStyles) is in the documentation.

You can use this as:
```
chatElementRef.messageStyles = {
    default: {
      shared: {
        bubble: {
          backgroundColor: 'yellow',
        },
      },
    },
    html: {
      bubble: {
        backgroundColor: 'purple',
      }
    }
  }
```

This sums up the current developments. I am still working on this and will let you know about further upgrades. Thanks!!

**Comment by OvidijusParsiunas** [comment]
I have fixed a few small bugs and added more functionality in version `9.0.14` of the dev packages.

By adding the class `deep-chat-suggestion-button` to your custom button, deep chat will automatically style it and add a custom event that sends a new message that contains the text of the button when the user clicks on it. This is one of the first preconfigured deep-chat elements that I will add in the future.

How to use the button:

```
// use similar syntax for when responding to a server request etc.

chatElementRef.initialMessages = [
    {
      role: 'user',
      html: '<button class="deep-chat-suggestion-button">How are you doing?</button>',
    }
  ]
```

<img width="329" alt="Screenshot 2023-10-01 at 22 15 58" src="https://github.com/OvidijusParsiunas/deep-chat/assets/18709577/beef4fba-e889-4838-8226-b38b8ba4491f">

Ofcourse the bubble background can be removed to make it a simple button using the `messageStyles` property mentioned above.

I am still working on this feature and will keep you posted on the latest. Thanks!

**Comment by OvidijusParsiunas** [comment]
I have updated the dev packages to version `9.0.15`.

This includes a new class that can be added to the custom element (must be on the parent-most element) called `deep-chat-temporary-element`. This will remove the element from the chat when a new message has been submitted, e.g. when referring to the suggestion button above, it will remove the element once it is clicked.

Example use:

```
chatElementRef.initialMessages = [
    {
      role: 'user',
      html: '<button class="deep-chat-suggestion-button deep-chat-temporary-element">How are you doing?</button>',
    }
  ]
```

This class will also not trigger the [onNewMessage](https://deepchat.dev/docs/events#onNewMessage) event and the message will not be returned in the [getMessages](https://deepchat.dev/docs/methods#getMessages) method as it is a temporary message.

**Comment by madhushan-t** [comment]
@OvidijusParsiunas Testing this out for the first time in Angular framework and it works great. Here's a first attempt:

<img width="391" alt="Screenshot 2023-10-02 at 2 57 59 PM" src="https://github.com/OvidijusParsiunas/deep-chat/assets/145786403/46f6c898-dedb-4037-a496-2d75a4f5639c">

1. Curiously, I'm having some trouble replacing the html content with a custom component 🤔  Is that still in the works or is it out of scope for this enhancement?

2. Is there an implicit order of operations when consuming the various response types? I found that it prefers `text` over `html` if both were provided. Is that expected? Since only one of them should be provided anyway, this makes sense.

3. Fun fact: in addition to looking for the html field in the backend's API response, I'm using the `responseInterceptor` to build that html. E.g. an API would realistically return some generic content like:
```
"pluginType": "plotly",
"data": [{"x": ["a", "b", "c"], "y": [1, 2, 3]}]
```
Then, based on the `pluginType`, we could selectively update the `html` field to render the data as needed, which in the above case would be something like:

```
response.html = `<chart [data]="data"></chart>`
```
This would help to decouple the backend from the UI.

**Comment by OvidijusParsiunas** [comment]
Hi @madhushan-t, thankyou for testing it out. I am not beside my development computer at the moment so I'll try to answer the questions as best as I can:

1. When referring to the ability to add custom elements/components, this should work for [custom elements](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements) that are defined to be as such. What you are using is likely an Angular component and the problem with that is that Angular is in charge of rendering and controlling its states via modules, this is something that Deep Chat can't do because it simply infers the string using `messageElements.bubbleElement.innerHTML = '<custom-element></custom-element>';` and leaves the browser to do the rest of the job.
2. 
3.

**Comment by OvidijusParsiunas** [comment]
Hi @madhushan-t, thankyou for testing it out. I am not beside my development computer at the moment so I'll try to answer the questions as best as I can:

1. When referring to the ability to add custom elements/components, this should work for [custom elements](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements) that are defined to be as such. What you are using is likely an Angular component and the problem with that is that Angular is in charge of rendering and controlling its states via modules, this is something that Deep Chat can't do because it simply infers the string using `messageElements.bubbleElement.innerHTML = '<custom-element></custom-element>';` and leaves the browser to do the rest of the job.
Having said that, most popular frameworks have the ability to turn their components into custom elements and I know this can be done in Angular (using [view encapsulation](https://angular.io/guide/view-encapsulation) if I remember correctly). I'll see if I can find out a way to help you do this tomorrow.

2. (Your message does not seem to include 2)

3. Currently it is either `html` or `text`, but I am very open to allowing and generating two messages when both are defined. I'll try to experiment with this tomorrow to see if there would be any problems with this.

4. Interceptors are definitely a great way to apply the `html`! I will leave the final decision of how to format the response to  the developers themselves to suit their preferences. Thankyou for a really nice example though!

I will do some testing around the things I talked about above tomorrow to see if I can improve them. Will keep you posted!

**Comment by OvidijusParsiunas** [comment]
I have updated the dev versions to `9.0.16`. This will now allow `text`, `files` and `html` to be in the same message (displayed in that order). If you also define `error` - they will not be displayed and instead an error message will be shown.
The `deep-chat-temporary-element` class has been renamed to `deep-chat-temporary-message` for better clarity.

I have additionally tested this solution in Angular, and as I have mentioned you will need to use the [viewEncapsulation](https://angular.io/guide/view-encapsulation) property for Angular components. This converts your component into a custom-element/shadow-element, and after you add it to the custom element registry (`customElements.define`) you can use it by following code examples above.

Here is a live example for this:

https://codesandbox.io/s/deep-chat-angular-custom-element-f2snlv?file=/src/app/angular-custom-element.ts
(Check the `angular-custom-element.ts` and `app.module.ts` files for the important code)

It is important to note that once inside the chat component, the Angular component is no longer treated as if it is owned by Angular, so when you want to pass values into it - you need to pass them as html attributes using the `attribute="value"` syntax. A nice explanation on what attributes are can be found [here](https://stackoverflow.com/questions/6003819/what-is-the-difference-between-properties-and-attributes-in-html).

The example above is just one small example of how you can create custom elements. Angular and other frameworks have multiple ways of creating custom elements and you can export them from their projects and inject them into your application just like `deep-chat` and use them in the `html` property.
(Yes, you can include Deep Chat inside Deep Chat with `html: "<deep-chat></deep-chat>"`. Pretty powerful, right? 😄)

Let me know if this works for you, and I'll keep updating.

**Comment by OvidijusParsiunas** [comment]
Small question, the `deep-chat-suggestion-button` class helps create buttons automatically, do you see any other common elements that people may want. I can try to add  more for this release or in the future.

Thanks!

**Comment by madhushan-t** [comment]
@OvidijusParsiunas Good question. A few ideas besides the suggestion button...
1) a simple approval dialog in situations where you may want a human-in-the-loop interaction. It would just trigger a yes / no text-prompt upon clicking. Or maybe even allow the positive and negative responses to be configured by the developer. 

2) a more general implementation of the suggestion button, which would take a list of suggestions:
<img width="362" alt="Screenshot 2023-10-03 at 2 09 13 PM" src="https://github.com/OvidijusParsiunas/deep-chat/assets/145786403/c0f67b2b-dec9-4a15-a593-ac678a33e9d9">

I can't think of anything specific right now, but I know this enhancement you've added will allow for a tonne of custom use-cases!

**Comment by OvidijusParsiunas** [comment]
Thankyou for the suggestions. I think all of those things can be achieved with the current API but I will add examples for your mentioned specific cases in the documentation as they are very valid! Thanks again!

**Comment by OvidijusParsiunas** [comment]
I have realised that allowing custom elements inside messages allows users to create messages that have varying dimensions, e.g. message content that autopopulates over time thus increasing the message height.

In the light of this, I have added a new method called `scrollToBottom` which does what the name says, scrolls the chat to bottom. This should help out in cases where the dynamic element full height is reached over-time. This will probably also be useful for developers working on other stuff too.
This is now present in the dev packages version `9.0.17`.

**Comment by madhushan-t** [comment]
@OvidijusParsiunas Trying this out right now 💯 
That codesandbox link you shared doesn't seem to be working for me 🤔

**Comment by OvidijusParsiunas** [comment]
What kind of error are you getting? Also can you try to hit the refresh button above the preview to see if it helps.

<img width="243" alt="image" src="https://github.com/OvidijusParsiunas/deep-chat/assets/18709577/0f3e4212-704a-4f29-b211-9fe0df5133a7">

**Comment by madhushan-t** [comment]
Your deployed app shows up great, but when navigating to the sandbox, I see this:

<img width="1728" alt="Screenshot 2023-10-05 at 9 40 50 PM" src="https://github.com/OvidijusParsiunas/deep-chat/assets/145786403/75d2c15f-3254-4182-82cb-0ec8dc36b155">

I'll keep digging!

**Comment by OvidijusParsiunas** [comment]
Hey, thankyou for more info. I have tested the link on other devices and it seems to work fine for me. It may be something to do with your device/internet/proxy, I am not quite sure.
Could you perhaps try to see if our standard [Angular example](https://codesandbox.io/s/deep-chat-angular-mk2v62?file=/src/app/app.component.html) (from the [Frameworks](https://deepchat.dev/examples/frameworks) page) loads for you as the sandbox you are having problems with is a fork from the exemplar one. Thanks!

**Comment by OvidijusParsiunas** [comment]
I have also updated the dev packages to version `1.0.19` which has now changed the behaviour of the `deep-chat-suggestion-button` class to only perform its actions and the devs will also need to add the `deep-chat-button` class to apply the styling. The reason for this decision was to allow users to apply default styling without having the suggestion button logic also applied with it.

I am also currently changing the `MessageStyles` object's `html` property to use `MessageRoleStyles` type instead of `MessageElementsStyles` as users may want different different bubble styling applied depending on the role.

Due to all of these changes, and a ton of documentation that will need to be changed, it is likely that the release will need to be pushed out to Sunday/Monday or later until I am fully confident that the API is solid.

Thanks!

**Comment by madhushan-t** [comment]
@OvidijusParsiunas Thanks for the update. As for the codesandbox issue, you're right it appears to have been an issue on my device. Worked perfectly on another machine. Thanks for the support!

Looking forward to the upcoming release 🚀 

Best,
Madhu

**Comment by OvidijusParsiunas** [comment]
Good to hear it works!

The `dev` packages has been updated to version `9.0.20` which contains an optimization for custom elements styles. Previously if you defined an element such as `<button style="background-color: green" class="deep-chat-button">Hi</button>`, the `"deep-chat-button`" class would have overwriten the `style="background-color: green"` which is technically incorrect as it does not conform to the css style declaration precedence rules, hence now styles inside `style=` are preserved.

**Comment by OvidijusParsiunas** [comment]
Hey, the `dev` packages have been updated to version `9.0.21` which has the `html` property's type in [`messageStyles`](https://deepchat.dev/docs/messageStyles#messageStyles) to [MessageRoleStyles](https://deepchat.dev/docs/messageStyles#MessageRoleStyles). Additionally this contains a bug fix for [`initialMessages`](https://deepchat.dev/docs/messages#initialMessages) when a file is used.

**Comment by OvidijusParsiunas** [comment]
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:
We finally did it! The custom elements feature has now finally been released to the main packages in version `1.3.11`! Check it out in [messages HTML](https://deepchat.dev/docs/messages/HTML) section of the documentation.
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

Thankyou very much for raising this feature, being patient and sticking through it all the way up until now!
I will be closing the issue, but the work for this feature will ofcourse keep going as the component is improved daily.
If you experience any issues or have suggestions for improvements regarding anything, you are always welcome to raise new issues! Thanks!
